src/requestor/routes/query.py

```diff
@@ -145,19 +153,43 @@ async def check_user_resource_paths(
 
     token_claims = await auth.get_token_claims()
     username = token_claims["context"]["user"]["name"]
-
+    if not permissions:
+        permissions = ["reader", "storage-reader"]
     res = {}
     user_requests = await get_user_requests(username)
-    for resource_path in resource_paths:
-        requests = [
-            r
-            for r in user_requests
-            if r.status not in config["DRAFT_STATUSES"]
-            and r.status not in config["FINAL_STATUSES"]
-            # TODO update logic to handle `policy_id` (PXP-8829)
-            and arborist.is_path_prefix_of_path(r.resource_path, resource_path)
+    positive_requests = [
+        r
+        for r in user_requests
+        if not r.revoke
+        and r.status not in config["DRAFT_STATUSES"]
+        and r.status not in config["FINAL_STATUSES"]
+    ]
+    existing_policies = await arborist.list_policies(
+        api_request.app.arborist_client, expand=True
+    )
+    # Initiate everything to False
+    res = {r: False for r in resource_paths}
+    for r in positive_requests:
+        # Get the policy
+        policy = get_policy_for_id(r.policy_id, existing_policies)
+        if policy is None:
+            continue
+        # Flatten permissions
+        policy_permission_ids = [
+            permission["id"]
+            for role in policy["roles"]
+            for permission in role["permissions"]
         ]
-        res[resource_path] = len(requests) > 0
+        # Continue to next request if all permissions in the request are not present in the policy
+        if not all(permission in policy_permission_ids for permission in permissions):
+            continue
+        # find if a resource path matches
+        for rp in policy["resource_paths"]:
+            for resource_path in resource_paths:
+                if arborist.is_path_prefix_of_path(rp, resource_path):
+                    # update res dictionary
+                    res[f"{resource_path}"] = True
+
```

Comment:
we should break out of the 2 `for` loops after setting the result to True

Reply:
I didn't think so! If let us say the policy has a `resource_path --> /records` and the `resource_paths` sent by user has `[/records/audio, /records/video]` then it makes sense to not break the loop right?

Correct me if I understood this incorrectly. :D

Reply:
oh sorry, you're right

a small improvement might be to do `for resource_path in <resource_paths that haven't been found yet>` on line 188. Not sure if it's worth it but the portal sends a long list of resource paths (one for each item in the study viewer), hence the concern